File: client/typescript/gamium-client-common/src/actions/action-chain.ts
import {
  ActionParam,
  ActionsParamT,
  AppQuitParamT,
  ErrorCode,
  InputKeyParamT,
  InputKeyPressType,
  InputMouseButtonCode,
  InputMouseParamT,
  InputMousePressType,
  InputSetTextParamT,
  MovePlayerParamT,
  ObjectLocatorBy,
  SleepParamT,
  Vector2T,
  Vector3T,
} from "../protocols/generated";
import * as gl from "gl-matrix";
import { GamiumError } from "../errors/gamium-error";
import { createActions, GamiumService } from "../gamium-service";
import { Locator } from "../locator/locator";
import {
  ActionsClickOptions,
  DefaultActionClickOptions,
} from "../options/actions-click-options";
import {
  ActionsDragOptions,
  DefaultActionsDragOptions,
} from "../options/actions-drag-options";
import {
  ActionsMoveOptions,
  DefaultActionMoveOptions,
} from "../options/actions-move-options";
import { DefaultActionScrollOptions } from "../options/actions-scroll-options";
import {
  DefaultMovePlayerOptions,
  MovePlayerOptions,
} from "../options/move-player-options";
import {
  DefaultSendKeyOptions,
  SendKeyOptions,
} from "../options/send-key-options";
import {
  DefaultSetTextOptions,
  SetTextOptions,
} from "../options/set-text-options";
import { ActionResult, Vector2, Vector3 } from "../protocols/types";
import { tryify, TryResult } from "../try";
import { Plain } from "../types";
import { KeyBy } from "./key-by";

const _dummyParam = new ActionsParamT();
type ActionParamTypes = NonNullable<(typeof _dummyParam.actions)[0]>;

interface ActionPacketTypes {
  paramEnum: ActionParam;
  param: ActionParamTypes;
}

export const createSleep = (param: Plain<SleepParamT>): ActionPacketTypes => {
  return {
    paramEnum: ActionParam.Actions_SleepParam,
    param: new SleepParamT(param.ms),
  };
};

export const createInputKey = (
  param: Plain<InputKeyParamT>
): ActionPacketTypes => {
  return {
    paramEnum: ActionParam.Actions_InputKeyParam,
    param: new InputKeyParamT(param.press, param.codes),
  };
};

export const createInputMouse = (
  param: Plain<InputMouseParamT>
): ActionPacketTypes => {
  return {
    paramEnum: ActionParam.Actions_InputMouseParam,
    param: new InputMouseParamT(
      param.press,
      param.button,
      param.position,
      param.delta
    ),
  };
};

export const createInputSetText = (
  param: Plain<InputSetTextParamT>
): ActionPacketTypes => {
  return {
    paramEnum: ActionParam.Actions_InputSetTextParam,
    param: new InputSetTextParamT(param.objectId, param.text),
  };
};

export const createMovePlayer = (
  param: Plain<MovePlayerParamT>
): ActionPacketTypes => {
  return {
    paramEnum: ActionParam.Actions_MovePlayerParam,
    param: new MovePlayerParamT(
      param.playerObjectId,
      param.cameraObjectId,
      param.position,
      param.by,
      param.epsilon,
      param.checkHeight
    ),
  };
};

export const createAppQuit = (
  param: Plain<AppQuitParamT>
): ActionPacketTypes => {
  return {
    paramEnum: ActionParam.Actions_AppQuitParam,
    param: new AppQuitParamT(param.exitCode, param.delayMs),
  };
};

export class ActionChain {
  constructor(
    private gamiumService: GamiumService,
    private actions: ActionPacketTypes[] = []
  ) {}

  private addAction(action: ActionPacketTypes): ActionChain {
    this.actions.push(action);
    return this;
  }

  sleep(millseconds: number): ActionChain {
    this.addAction(createSleep({ ms: millseconds }));
    return this;
  }

  click(
    position: Vector2,
    options: Partial<ActionsClickOptions> = DefaultActionClickOptions()
  ): ActionChain {
    const optionMixed = { ...DefaultActionClickOptions(), ...options };

    const vec2Pos = new Vector2T(position.x, position.y);
    this.addAction(
      createInputMouse({
        press: InputMousePressType.DOWN,
        button: InputMouseButtonCode.LEFT,
        position: vec2Pos,
        delta: new Vector2T(),
      })
    );
    this.addAction(createSleep({ ms: optionMixed.durationMs }));
    this.addAction(
      createInputMouse({
        press: InputMousePressType.UP,
        button: InputMouseButtonCode.LEFT,
        position: vec2Pos,
        delta: new Vector2T(),
      })
    );
    this.addAction(createSleep({ ms: 33 }));
    return this;
  }

  move(
    position: Vector2,
    options: Partial<ActionsMoveOptions> = DefaultActionMoveOptions()
  ): ActionChain {
    const optionMixed = { ...DefaultActionMoveOptions(), ...options };

    const vec2Pos = new Vector2T(position.x, position.y);
    this.addAction(
      createInputMouse({
        press: InputMousePressType.MOVE,
        button: InputMouseButtonCode.LEFT,
        position: vec2Pos,
        delta: new Vector2T(),
      })
    );
    return this;
  }

  drag(
    fromPosition: Vector2,
    toPosition: Vector2,
    options: Partial<ActionsDragOptions> = DefaultActionsDragOptions()
  ): ActionChain {
    const optionMixed = { ...DefaultActionsDragOptions(), ...options };
    const fromVec2 = new Vector2T(fromPosition.x, fromPosition.y);
    const toVec2 = new Vector2T(toPosition.x, toPosition.y);

    this.addAction(
      createInputMouse({
        press: InputMousePressType.DOWN,
        button: InputMouseButtonCode.LEFT,
        position: fromVec2,
        delta: new Vector2T(),
      })
    );

    const deltaCount = optionMixed.durationMs / (optionMixed.intervalMs + 1);
    const deltaVector = gl.vec2.div(
      gl.vec2.create(),
      gl.vec2.sub(
        gl.vec2.create(),
        [toVec2.x, toVec2.y],
        [fromVec2.x, fromVec2.y]
      ),
      [deltaCount, deltaCount]
    );
    this.addAction(createSleep({ ms: optionMixed.intervalMs }));
    for (let i = 1; i < deltaCount - 1; i++) {
      const nextPosition = new Vector2T(
        fromVec2.x + deltaVector[0] * i,
        fromVec2.y + deltaVector[1] * i
      );
      this.addAction(
        createInputMouse({
          press: InputMousePressType.MOVE,
          button: InputMouseButtonCode.LEFT,
          position: nextPosition,
          delta: new Vector2T(),
        })
      );
      this.addAction(createSleep({ ms: optionMixed.intervalMs }));
    }
    this.addAction(
      createInputMouse({
        press: InputMousePressType.UP,
        button: InputMouseButtonCode.LEFT,
        position: toVec2,
        delta: new Vector2T(),
      })
    );
    this.addAction(createSleep({ ms: optionMixed.intervalMs }));

    return this;
  }

  scroll(
    position: Vector2,
    delta: Vector2,
    options = DefaultActionScrollOptions()
  ): ActionChain {
    const vec3Pos = new Vector3T(position.x, position.y, 0);
    const vec2Delta = new Vector2T(delta.x, delta.y);
    this.addAction(
      createInputMouse({
        press: InputMousePressType.SCROLL,
        button: InputMouseButtonCode.LEFT,
        position: vec3Pos,
        delta: vec2Delta,
      })
    );
    this.addAction(createSleep({ ms: options.durationMs }));
    return this;
  }

  sendKeys(
    byList: KeyBy[],
    options: Partial<SendKeyOptions> = DefaultSendKeyOptions()
  ): ActionChain {
    const codes = byList.map((c) => c.str);

    const optionMixed: SendKeyOptions = {
      ...DefaultSendKeyOptions(),
      ...options,
    };
    this.addAction(
      createInputKey({ press: InputKeyPressType.DOWN, codes: codes })
    );
    this.addAction(createSleep({ ms: optionMixed.duratiomMs }));
    this.addAction(
      createInputKey({ press: InputKeyPressType.UP, codes: codes })
    );
    this.addAction(createSleep({ ms: 33 }));
    return this;
  }

  setText(
    locator: Locator,
    text: string,
    options: Partial<SetTextOptions> = DefaultSetTextOptions()
  ): ActionChain {
    if (locator.by !== ObjectLocatorBy.Path) {
      throw new Error("setText only support Path locator");
    }
    this.addAction(createInputSetText({ objectId: locator.str, text: text }));
    this.addAction(createSleep({ ms: 33 }));
    return this;
  }

  movePlayer(
    playerLocator: Locator,
    cameraLocator: Locator,
    dest: Vector3,
    options: Partial<MovePlayerOptions> = DefaultMovePlayerOptions()
  ): ActionChain {
    if (playerLocator.by !== ObjectLocatorBy.Path) {
      throw new Error("movePlayer only support Path locator");
    }
    if (cameraLocator.by !== ObjectLocatorBy.Path) {
      throw new Error("movePlayer only support Path locator");
    }

    const optionsMixed: MovePlayerOptions = {
      ...DefaultMovePlayerOptions(),
      ...options,
    };
    this.addAction(
      createMovePlayer({
        playerObjectId: playerLocator.str,
        cameraObjectId: cameraLocator.str,
        position: new Vector3T(dest.x, dest.y, dest.z),
        by: optionsMixed.by,
        epsilon: optionsMixed.epsilon,
        checkHeight: optionsMixed.checkHeight,
      })
    );
    return this;
  }

  appQuit(exitCode = 0, delayMs = 10): ActionChain {
    this.addAction(createAppQuit({ exitCode: exitCode, delayMs: delayMs }));
    return this;
  }

  async perform(): Promise<ActionResult[]> {
    if (this.actions.length === 0) {
      throw new GamiumError(
        ErrorCode.InvalidParameter,
        "ActionChains.perform actions is empty"
      );
    }
    const actionTypes = this.actions.map((action) => {
      return action.paramEnum;
    });
    const actions = this.actions.map((action) => {
      return action.param;
    });
    const res = await this.gamiumService.request(
      createActions({ actionsType: actionTypes, actions: actions })
    );
    const results = res.results.map((result) => {
      const actionResult: ActionResult = {
        error: {
          code: result.error?.code ?? ErrorCode.None,
          reason: result.error?.reason?.toString() ?? "",
        },
      };
      return actionResult;
    });
    this.checkActionReturn(results);

    return results;
  }

  async tryPerform(): Promise<TryResult<ActionResult[]>> {
    return tryify(this.perform());
  }

  private checkActionReturn(
    res: Awaited<ReturnType<ActionChain["perform"]>>
  ): void {
    for (let i = 0; i < res.length; i++) {
      const actRes = res[i];
      if (actRes === undefined) {
        throw new GamiumError(
          ErrorCode.InternalError,
          `ActionChains.perform action ${i} return undefined`
        );
      }
      if (!actRes.error) {
        throw new GamiumError(
          ErrorCode.InternalError,
          "pressKey action result error is null"
        );
      }
      if (actRes.error.code !== ErrorCode.None) {
        throw new GamiumError(
          ErrorCode.InternalError,
          `pressKey action[${i}] result error code:${String(
            actRes.error
          )}}, msg:${String(actRes.error.reason)}`
        );
      }
    }
  }
}

File: client/python/gamium/gamium/actions/action_chain.py
from typing import Generic, List
import flatbuffers
from gamium.protocol.generated.Packets.ActionParam import ActionParam
from gamium.protocol.generated.Types.InputMousePressType import InputMousePressType
from gamium.protocol.generated.Types.ErrorCode import ErrorCode
from gamium.protocol.generated.Types.InputKeyPressType import InputKeyPressType
from gamium.protocol.generated.Types.InputMouseButtonCode import (
    InputMouseButtonCode,
)
from gamium.protocol.generated.Types.ObjectLocatorBy import ObjectLocatorBy
from gamium.protocol.generated.Packets.ActionParamSingle import (
    ActionParamSingleT,
)
from gamium.protocol.generated.Packets.ActionResult import ActionResultT
from gamium.protocol.generated.Packets.Actions.AppQuitParam import (
    AppQuitParamT,
)
from gamium.protocol.generated.Packets.Actions.InputKeyParam import (
    InputKeyParamT,
)
from gamium.protocol.generated.Packets.Actions.InputMouseParam import (
    InputMouseParamT,
)
from gamium.protocol.generated.Packets.Actions.InputSetTextParam import (
    InputSetTextParamT,
)
from gamium.protocol.generated.Packets.Actions.MovePlayerParam import (
    MovePlayerParamT,
)
from gamium.protocol.generated.Packets.Actions.SleepParam import SleepParamT
from gamium.protocol.generated.Types.Vector2 import Vector2T
from gamium.protocol.generated.Types.Vector3 import Vector3T
from gamium.igamium_service import IGamiumService, create_actions
from gamium.options import (
    ActionClickOptions,
    ActionDragOptions,
    ActionMoveOptions,
    ActionScrollOptions,
    MovePlayerOptions,
    SendKeyOptions,
    SetTextOptions,
)
from gamium.actions.key_by import KeyBy
from gamium.errors.gamium_error import GamiumError
from gamium.locator.locator import Locator
from gamium.protocol.types import Vector2, Vector3
from gamium.utils.generics import P


class ActionPacketTypes(Generic[P]):
    def __init__(self, param_type: int, param: P):
        self.param_type = param_type  # ActionParam
        self.param: P = param


def create_sleep(ms: int) -> ActionPacketTypes[SleepParamT]:
    param = SleepParamT()
    param.ms = ms
    return ActionPacketTypes(ActionParam.Actions_SleepParam, param)


def create_input_key(press: int, codess: List[str]) -> ActionPacketTypes[InputKeyParamT]:
    param = InputKeyParamT()
    param.press = press
    param.codes = codess
    return ActionPacketTypes(ActionParam.Actions_InputKeyParam, param)


def create_input_mouse(press: int, button: int, position: Vector2T, delta: Vector2T) -> ActionPacketTypes[InputMouseParamT]:
    param = InputMouseParamT()
    param.press = press
    param.button = button
    param.position = position
    param.delta = delta
    return ActionPacketTypes(ActionParam.Actions_InputMouseParam, param)


def create_set_text(object_id: str, text: str) -> ActionPacketTypes[InputSetTextParamT]:
    param = InputSetTextParamT()
    param.objectId = object_id
    param.text = text
    return ActionPacketTypes(ActionParam.Actions_InputSetTextParam, param)


def create_move_player(
    player_object_id: str,
    camera_object_id: str,
    position: Vector3T,
    by: int,
    epsilon: float,
    check_height: bool,
) -> ActionPacketTypes[MovePlayerParamT]:
    param = MovePlayerParamT()
    param.playerObjectId = player_object_id
    param.cameraObjectId = camera_object_id
    param.position = position
    param.by = by
    param.epsilon = epsilon
    param.checkHeight = check_height
    return ActionPacketTypes(ActionParam.Actions_MovePlayerParam, param)


def create_app_quit(exit_code: int, delay_ms: int) -> ActionPacketTypes[AppQuitParamT]:
    param = AppQuitParamT()
    param.exitCode = exit_code
    param.delayMs = delay_ms
    return ActionPacketTypes(ActionParam.Actions_AppQuitParam, param)


class ActionChain:
    def __init__(self, service: IGamiumService) -> None:
        self._service = service
        self.actions: List[ActionPacketTypes] = []

    def __add_action(self, action: ActionPacketTypes):
        self.actions.append(action)
        return self

    def sleep(self, ms: int):
        return self.__add_action(create_sleep(ms))

    def click(
        self,
        position: Vector2,
        options: ActionClickOptions = ActionClickOptions(),
    ):
        fb_position = Vector2T()
        fb_position.x = position.x
        fb_position.y = position.y

        self.__add_action(
            create_input_mouse(
                InputMousePressType.DOWN,
                InputMouseButtonCode.LEFT,
                fb_position,
                Vector2T(),
            )
        )
        self.__add_action(create_sleep(options.duration_ms))
        self.__add_action(create_input_mouse(InputKeyPressType.UP, InputMouseButtonCode.LEFT, fb_position, Vector2T()))
        self.__add_action(create_sleep(33))
        return self

    def move(
        self,
        position: Vector2,
        options: ActionMoveOptions = ActionMoveOptions(),
    ):
        fb_position = Vector2T()
        fb_position.x = position.x
        fb_position.y = position.y

        self.__add_action(
            create_input_mouse(
                InputMousePressType.MOVE,
                InputMouseButtonCode.LEFT,
                fb_position,
                Vector2T(),
            )
        )
        return self

    def drag(
        self,
        from_position: Vector2,
        to_position: Vector2,
        options: ActionDragOptions = ActionDragOptions(),
    ):
        fb_from_position = Vector2T()
        fb_from_position.x = from_position.x
        fb_from_position.y = from_position.y

        fb_to_position = Vector2T()
        fb_to_position.x = to_position.x
        fb_to_position.y = to_position.y

        self.__add_action(
            create_input_mouse(
                InputMousePressType.DOWN,
                InputMouseButtonCode.LEFT,
                fb_from_position,
                Vector2T(),
            )
        )

        delta_count = int(options.duration_ms / (options.interval_ms + 1))
        delta_vector = [
            (to_position.x - from_position.x) / delta_count,
            (to_position.y - from_position.y) / delta_count,
        ]

        for i in range(delta_count - 1):
            fb_next_position = Vector2T()
            fb_next_position.x = from_position.x + delta_vector[0] * i
            fb_next_position.y = from_position.y + delta_vector[1] * i
            self.__add_action(
                create_input_mouse(
                    InputMousePressType.MOVE,
                    InputMouseButtonCode.LEFT,
                    fb_next_position,
                    Vector2T(),
                )
            )
            self.__add_action(create_sleep(options.interval_ms))

        self.__add_action(
            create_input_mouse(
                InputMousePressType.UP,
                InputMouseButtonCode.LEFT,
                fb_to_position,
                Vector2T(),
            )
        )
        return self

    def scroll(
        self,
        position: Vector2,
        delta: Vector2,
        options: ActionScrollOptions = ActionScrollOptions(),
    ):
        fb_position = Vector2T()
        fb_position.x = position.x
        fb_position.y = position.y

        fb_delta = Vector2T()
        fb_delta.x = delta.x
        fb_delta.y = delta.y

        self.__add_action(
            create_input_mouse(
                InputMousePressType.SCROLL,
                InputMouseButtonCode.LEFT,
                fb_position,
                fb_delta,
            )
        )
        self.__add_action(create_sleep(options.duration_ms))
        return self

    def send_keys(self, by_list: List[KeyBy], options: SendKeyOptions = SendKeyOptions()):
        codes = []
        for by in by_list:
            codes.append(by.str)

        self.__add_action(create_input_key(InputKeyPressType.DOWN, codes))
        self.__add_action(create_sleep(options.duration_ms))
        self.__add_action(create_input_key(InputKeyPressType.UP, codes))
        self.__add_action(create_sleep(33))

        return self

    def set_text(
        self,
        locator: Locator,
        text: str,
        options: SetTextOptions = SetTextOptions(),
    ):
        if locator.by != ObjectLocatorBy.Path:
            raise GamiumError(ErrorCode.InvalidParameter, "setText only support Path locator")

        self.__add_action(create_set_text(locator.str, text))
        self.__add_action(create_sleep(33))
        return self

    def move_player(
        self,
        player_locator: Locator,
        camera_locator: Locator,
        dest: Vector3,
        options: MovePlayerOptions = MovePlayerOptions(),
    ):
        fb_dest = Vector3T()
        fb_dest.x = dest.x
        fb_dest.y = dest.y
        fb_dest.z = dest.z

        if player_locator.by != ObjectLocatorBy.Path or camera_locator.by != ObjectLocatorBy.Path:
            raise GamiumError(ErrorCode.InvalidParameter, "movePlayer only support Path locator")

        self.__add_action(
            create_move_player(
                player_locator.str,
                camera_locator.str,
                fb_dest,
                options.by,
                options.epsilon,
                options.check_height,
            )
        )
        self.__add_action(create_sleep(33))
        return self

    def app_quit(self, exit_code: int = 0, delay_ms: int = 10):
        self.__add_action(create_app_quit(exit_code, delay_ms))
        return self

    def perform(self) -> List[ActionResultT]:
        if len(self.actions) == 0:
            raise GamiumError(
                ErrorCode.InvalidParameter,
                "ActionChains.perform. actions shoult not be empty",
            )

        actions_array: List[str] = []
        for action in self.actions:
            action_param = ActionParamSingleT()
            action_param.actionType = action.param_type
            action_param.action = action.param

            builder = flatbuffers.Builder()
            request_offset = action_param.Pack(builder)
            builder.Finish(request_offset)

            param_bytes = builder.Output()
            param_str = "".join(chr(byte) for byte in param_bytes)
            actions_array.append(param_str)

        res = self._service.request(create_actions(actions_array))
        self.__check_action_return(res.results)
        return res.results

    def __check_action_return(self, results: List[ActionResultT]):
        for i in range(len(results)):
            result = results[i]
            if result is None:
                raise GamiumError(
                    ErrorCode.InternalError,
                    f"ActionChains.perform action {i} return undefined",
                )

            if result.error is None:
                raise GamiumError(
                    ErrorCode.InternalError,
                    "ActionChains.perform. action failed. error is None",
                )
            if result.error.code != ErrorCode.None_:
                raise GamiumError(
                    ErrorCode.InternalError,
                    "ActionChains.perform. action failed. error code: " + str(result.error.code) + " error message: " + result.error.reason,
                )
